controllers/user/coursesProgressUser.js

```diff
@@ -0,0 +1,122 @@
+const User = require("../../models/User");
+const Course = require("../../models/Courses");
+const Resource = require("../../models/Resources");
+
+const CourseProgress = async (req, res) => {
+  const { userId, courseId, topicId, resourceId } = req.body;
+
+  try {
+    // Find the user by ID and populate CourseProgress and course
+    const user = await User.findById(userId).populate({
+      path: "CourseProgress.course",
+      populate: {
+        path: "lastViewedTopic",
+      },
+    });
+
+    if (!user) {
+      return res.status(404).json({ message: "User not found" });
+    }
+
+    if (!Array.isArray(user.CourseProgress)) {
+      return res
+        .status(500)
+        .json({ message: "Invalid user course progress structure" });
+    }
+
+    // Find the user's course progress
+    let courseProgress = user.CourseProgress.find(
+      (cp) => cp.course && cp.course.equals(courseId)
+    );
+
+    // Find the course by ID
+    const course = await Course.findById(courseId);
+    if (!course) {
+      return res.status(404).json({ message: "Course not found" });
+    }
+
+    // Find the position of the current topicId in the course topics list
+    const currentTopicIndex = course.topic.findIndex(
+      (t) => t._id.toString() === topicId
+    );
+    if (currentTopicIndex === -1) {
+      return res.status(404).json({ message: "Topic not found in course" });
+    }
+
+    if (!courseProgress) {
+      courseProgress = {
+        course: courseId,
+        lastViewedTopic: {
+          topicId: topicId,
+          lastViewedResource: resourceId,
+        },
+      };
+
+      user.CourseProgress.push(courseProgress);
+    } else {
+      // Display the lastViewedTopic stored in the user's progress
+
```

Comment:
Could you please delete this blank line?

Reply:
line delete